### PR TITLE
Fix segmentation fault in udp proxy filter

### PR DESCRIPTION
**Pull request by IssaAbuKalbein**
Commit Message: Fix segmentation fault in udp proxy filter
Additional Description: - proxy_filter.cc => Check host_info in onLoadDnsCacheComplete to verify that DNS lookup wasn't 
                                         failed (because in case of failure "read_callbacks_->continueFilterChain();" will remove the session).
                                      - udp_proxy_filter.cc => Add the session before calling "onNewSession" (because "onNewSession" will 
                                         remove the session if it failed to create upstream connection pool).
Risk Level: low, fix a bug
Testing: Unit Test
Docs Changes: None
Release Notes: None
Platform Specific Features: None

### CODE REVIEW

**Comment by IssaAbuKalbein** [comment]
\retest

File: source/extensions/filters/udp/udp_proxy/session_filters/dynamic_forward_proxy/proxy_filter.cc [hunk 1]
```diff
@@ -106,18 +106,23 @@ ReadFilterStatus ProxyFilter::onData(Network::UdpRecvData& data) {
...
-    datagrams_buffer_.pop();
-    read_callbacks_->injectDatagramToFilterChain(*buffered_datagram);
+  if (host_info && host_info->address()) {
+    while (!datagrams_buffer_.empty()) {
+      BufferedDatagramPtr buffered_datagram = std::move(datagrams_buffer_.front());
+      datagrams_buffer_.pop();
+      read_callbacks_->injectDatagramToFilterChain(*buffered_datagram);
+    }
+  } else {
+    ENVOY_LOG(debug, "DNS lookup failed");
```
Comment (RyanTheOptimist) [inline source/extensions/filters/udp/udp_proxy/session_filters/dynamic_forward_proxy/proxy_filter.cc:125]: I'm curious, if we simply LOG here (and don't explicitly set some sort of error condition) do we run the risk of getting ourselves in a state where we're just kinda stalled? Or does it all work out cleanly?

**Review by RyanTheOptimist — COMMENTED** [review]
Looks good, modulo a question

File: source/extensions/filters/udp/udp_proxy/session_filters/dynamic_forward_proxy/proxy_filter.cc (hunk 1 above)
Comment (IssaAbuKalbein) [reply source/extensions/filters/udp/udp_proxy/session_filters/dynamic_forward_proxy/proxy_filter.cc]: @RyanTheOptimist Changed the approach for the fix, can you take a look, please?

**Review by RyanTheOptimist — APPROVED** [review]

**Comment by RyanTheOptimist** [comment]
Looks like we need more test coverage:

```
Code coverage for source/extensions/filters/udp/udp_proxy is lower than limit of 96.6 (96.4)
```

https://storage.googleapis.com/envoy-pr/95b6021/coverage/source/extensions/filters/udp/udp_proxy/index.html
/wait

**Comment by RyanTheOptimist** [comment]
Looks like a TSAN failure:

**Comment by RyanTheOptimist** [comment]
Argh! Pressed the wrong button somehow.

**Comment by IssaAbuKalbein** [comment]
@RyanTheOptimist Fixed, can you take a look, please?

**Review by RyanTheOptimist — APPROVED** [review]

**Pull request merged**